R/round_coordinates.R

```diff
@@ -0,0 +1,140 @@
+#' This function can be used before publishing data in order to protect
+#' sensitive species and/or prevent theft of active cameras.
+#'
+#' @inheritParams check_camtrapdp
+#' @param digits Number of decimal places to round coordinates to (`1`,
+#' `2` or `3`).
+#' @return `x` with rounded coordinates as well as updated
+#'   `coordinateUncertainty` (in deployments) and `coordinatePrecision` (in
+#'   metadata).
+#' @family transformation functions
+#' @export
+#' @section Details:
+#' Rounding coordinates is a recommended method to generalize sensitive
+#' biodiversity information (see
+#' [Section 4.2](https://doi.org/10.15468/doc-5jp4-5g10#s-generalization)
+#' in Chapman 2020).
+#' Choose a number of digits that aligns with the sensitivity of the data and
+#' notice the effect on precision and uncertainty.
+#' Publish the coordinates as is (i.e. do not use this function) if the data
+#' are not sensitive.
+#'
+#' sensitivity | digits | coordinatePrecision
+#' --- | --- | ---
+#' high | 1 | 0.1
+#' medium | 2 | 0.01
+#' low | 3 | 0.001
+#'
+#' For records with `coordinateUncertainty = NA` the function will assume the
+#' coordinates were obtained by GPS and use `30 m` as original uncertainty,
+#' before adding uncertainty caused by rounding.
+#' The added rounding uncertainty is dependent on latitude and follows the
+#' categories in
+#' [Table 3](https://doi.org/10.15468/doc-gg7h-s853#table-uncertainty) in
+#' Chapman & Wieczorek 2020.
+#'
+#' latitude | 0.1 degree | 0.01 degree | 0.001 degree
+#' --- | --- | --- | ---
+#' 0° | 15691 m | 1570 m | 157 m
+#' 30° | 14697 m | 1470 m | 147 m
+#' 60° | 12461 m | 1246 m | 125 m
+#' 85° | 11211 m | 1121 m | 112 m
+#'
+#' @examples
+#' # Round coordinates of example package to 3 digits
+#' x <- example_dataset()
+#' x_rounded2 <- round_coordinates(x, 2)
+#'
+#' # coordinatePrecision is set in metadata
+#' x_rounded2$coordinatePrecision
+#'
+#' # coordinateUncertainty is set in data: original uncertainty (or 30) + 1470 m
+#' x_rounded2$data$deployments$coordinateUncertainty
+round_coordinates <- function(x, digits = 3) {
+  if (!(digits %in% c(1, 2, 3))) {
+    cli::cli_abort(
+      c("`digits` must be NULL, 1, 2 or 3."),
+      class = "camtrapdp_error_digits"
+    )
+  }
+
+  # Detect original number of digits from coordinatePrecision or data
+  original_precision <- x$coordinatePrecision
+  if (!is.null(original_precision)) {
+    original_digits <- -log10(original_precision) # 0.001 -> 3
+    if (digits >= original_digits) {
+      cli::cli_abort(
+        c(
+          "Can't round from {original_digits} to {digits} digits. ",
+          "`{original_digits}` is derived from the ",
+          "`x$coordinatePrecision={original_precision}`."
+        ),
+        class = "camtrapdp_error_precision"
+      )
+    }
+  } else {
+    original_digits <-
+      deployments(x) %>%
+      dplyr::mutate(
+        lat_digits = nchar(stringr::str_extract(.data$latitude, "\\d+$"))
```

Comment:
Is it possible to do this step without stringr or any other dependency? grep or alike perhaps?

Reply:
I personally like creating a helper for this: 


https://gist.github.com/PietrH/47701b02cb78a26eab4a25fdac8f17ef

Reply:
If you do, I would call it decimal_places or alike and also work in the scientific notation edge case, see https://stackoverflow.com/questions/5173692/how-to-return-number-of-decimal-places-in-r

Reply:
I replaced the stringr dependency with gsub(). But in this case, a latitude of 2.000 is having 3 digits. Do you want this to be 0 like in the link?

Reply:
Is that because you are counting the number of characters in the terminal digit only string? `000`?

Converting to a number and using log is also an option

Reply:
`2.000` being 3 digits is fine to me.

Reply:
@peterdesmet Can you give an example of a latitude in a camtrap DP package in scientific notation? I don't see how that would happen as it should be a number between -90 and 90.

Reply:
You're right, no need to code for that.